six/modules/c++/cphd/source/CPHDXMLControl.cpp

```diff
@@ -35,1090 +41,2414 @@ typedef xml::lite::Element* XMLElem;
+        XMLElem linePolyXML = newElement(numLatLons == 2 ? "Line" : "Polygon",
+                                         geoInfoXML);
+        setAttribute(linePolyXML, "size", str::toString(numLatLons));
+
+        for (size_t ii = 0; ii < numLatLons; ++ii)
+        {
+            XMLElem v = mCommon.createLatLon(numLatLons == 2 ? "Endpoint" : "Vertex",
+                         geoInfo.geometryLatLon[ii], linePolyXML);
+            setAttribute(v, "index", str::toString(ii + 1));
+        }
     }
 
-    if (vp.toaParameters.get() != NULL)
+    if (!geoInfo.name.empty())
+        setAttribute(geoInfoXML, "name", geoInfo.name);
+
+    for (size_t ii = 0; ii < geoInfo.geoInfos.size(); ++ii)
     {
-        XMLElem toaParametersXML = newElement("TOAParameters", vectorParametersXML);
-        createInt("DeltaTOA0", vp.toaParameters->deltaTOA0, toaParametersXML);
-        createInt("TOA_SS", vp.toaParameters->toaSS, toaParametersXML);
+        toXML(*geoInfo.geoInfos[ii], geoInfoXML);
     }
 
-    return vectorParametersXML;
-}
-
-XMLElem CPHDXMLControl::areaLineDirectionParametersToXML(
-    const std::string& name,
-    const AreaDirectionParameters& adp,
-    XMLElem parent)
-{
-    XMLElem adpXML = newElement(name, parent);
-    mCommon.createVector3D("UVectECF", adp.unitVector, adpXML);
-    createDouble("LineSpacing", adp.spacing, adpXML);
-    createInt("NumLines", adp.elements, adpXML);
-    createInt("FirstLine", adp.first, adpXML);
-    return adpXML;
+    return geoInfoXML;
 }
 
-XMLElem CPHDXMLControl::areaSampleDirectionParametersToXML(
-    const std::string& name,
-    const AreaDirectionParameters& adp,
-    XMLElem parent)
+XMLElem CPHDXMLControl::toXML(const MatchInformation& matchInfo, XMLElem parent)
 {
-    XMLElem adpXML = newElement(name, parent);
-    mCommon.createVector3D("UVectECF", adp.unitVector, adpXML);
-    createDouble("SampleSpacing", adp.spacing, adpXML);
-    createInt("NumSamples", adp.elements, adpXML);
-    createInt("FirstSample", adp.first, adpXML);
-    return adpXML;
+    return mCommon.convertMatchInformationToXML(&matchInfo, parent);
 }
 
-std::auto_ptr<Metadata> CPHDXMLControl::fromXML(const std::string& xmlString)
+/*
+ * FROM XML
+ */
+std::auto_ptr<Metadata> CPHDXMLControl::fromXML(
+        const std::string& xmlString,
+        const std::vector<std::string>& schemaPaths)
 {
     io::StringStream stringStream;
     stringStream.write(xmlString.c_str(), xmlString.size());
     xml::lite::MinidomParser parser;
     parser.parse(stringStream);
-    return fromXML(parser.getDocument());
+    return fromXML(parser.getDocument(), schemaPaths);
 }
 
-std::auto_ptr<Metadata> CPHDXMLControl::fromXML(const xml::lite::Document* doc)
+std::auto_ptr<Metadata> CPHDXMLControl::fromXML(
+        const xml::lite::Document* doc,
+        const std::vector<std::string>& schemaPaths)
 {
     std::auto_ptr<Metadata> cphd(new Metadata());
 
+    if(!schemaPaths.e
```

Comment:
For consistency, let's keep the names camel case (e.g. `codXmlVec`). The SICD XML parser often handles plural naming like `geoInfosXml`, but that may be silly in this case.